### PR TITLE
refactor: Extract common login code to script

### PR DESCRIPTION
**Pull request by ggrangel**
# Summary

Solves: #2974 

# Testing Process

- [ ] Add how to test.

## Checklist

- [x] Review the contributing [guidelines](https://github.com/slsa-framework/slsa-github-generator/blob/main/CONTRIBUTING.md)
- [x] Add a reference to related issues in the PR description.
- [ ] Update documentation if applicable.
- [ ] Add unit tests if applicable.
- [ ] Add changes to the [CHANGELOG](https://github.com/slsa-framework/slsa-github-generator/blob/main/CHANGELOG.md) if applicable.

### CODE REVIEW

**Comment by ggrangel** [comment]
I'm not exactly sure how to test this. 

I suspect I should use this [e2e script](https://github.com/slsa-framework/example-package/blob/main/.github/workflows/scripts/manual-run-e2e.sh) but I don't know how to point to my changes. 

Any help is appreciated.

File: .github/workflows/scripts/registry-login.sh [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+# 3. "my-project/my-image" -> "docker.io"
+# 4. "gcr.io" -> "gcr.io"
+if [[ "$UNTRUSTED_IMAGE_OR_REPO" == *"/"* ]]; then
+  # Extract possible domain (containing '.' or ':') as registry
+  untrusted_registry=$(echo "${UNTRUSTED_IMAGE_OR_REPO}" | cut -f1 -d "/" | { grep -E "\.|:" || true; })
+  if [ "${untrusted_registry}" == "" ]; then
+    # Use default registry if no domain is found
+    untrusted_registry="$DEFAULT_REGISTRY"
+  fi
+else
```
Comment (ramonpetgrave64) [inline .github/workflows/scripts/registry-login.sh:41]: IIUC, if your script can't parse the registry against the regex `*/*`, then fallback to the user-provided input.
This seems like different from the original behavior. I recommend re-using the original logic, so we can be more sure we're not regressing.

I know it's difficult to test workflows end-to-end on your own forks. The best we've had so far is to merge, and then wait for the e2e tests in slsa-framework/example-package to run. You might still be able to partially test this workflow on your own fork, however, after renaming a few references to point to your own fork.

Comment (ggrangel) [reply .github/workflows/scripts/registry-login.sh:32]: Thanks for the heads up. This approach is indeed safer.

File: .github/workflows/generator_container_slsa3.yml [hunk 2]
```diff
@@ -182,82 +180,20 @@ jobs:
...
-          # Here we get the first part and check if it has a '.' or ':'
-          # character in it to see if it's a domain name.
-          # See: https://stackoverflow.com/questions/37861791/how-are-docker-image-names-parsed#37867949
-          untrusted_registry="docker.io"
-          # NOTE: Do not fail the script if grep does not match.
-          maybe_domain=$(echo "${UNTRUSTED_IMAGE}" | cut -f1 -d "/" | { grep -E "\.|:" || true; })
-          if [ "${maybe_domain}" != "" ]; then
-            untrusted_registry="${maybe_domain}"
-          fi
+          ./scripts/registry-login.sh "${UNTRUSTED_IMAGE}" "${UNTRUSTED_SECRET_USERNAME-:${UNTRUSTED_INPUT_USERNAME}}" "${UNTRUSTED_PASSWORD}" "${GCP_ACCESS_TOKEN}"
```
Comment (ramonpetgrave64) [inline .github/workflows/generator_container_slsa3.yml:183]:
```suggestion
          ./scripts/registry-login.sh "${UNTRUSTED_IMAGE}" "${UNTRUSTED_SECRET_USERNAME:-${UNTRUSTED_INPUT_USERNAME}}" "${UNTRUSTED_PASSWORD}" "${GCP_ACCESS_TOKEN}"
```

File: .github/workflows/generator_container_slsa3.yml [hunk 3]
```diff
@@ -182,82 +180,20 @@ jobs:
...
-            password="${UNTRUSTED_PROVENANCE_PASSWORD}"
-            if [ "${password}" == "" ]; then
-                echo "provenance-registry-password is required." >&2
-                exit 1
-            fi
-          fi
-
-          echo "login to provenance registry: ${untrusted_provenance_domain}"
-          cosign login "${untrusted_provenance_domain}" -u "${username}" -p "${password}"
+          ./scripts/registry-login.sh "${UNTRUSTED_PROVENANCE_REPOSITORY}" "${UNTRUSTED_SECRET_USERNAME-:${UNTRUSTED_INPUT_USERNAME}}" "${UNTRUSTED_PASSWORD}" "${GCP_ACCESS_TOKEN}"
```
Comment (ramonpetgrave64) [inline .github/workflows/generator_container_slsa3.yml:196]:
```suggestion
          ./scripts/registry-login.sh "${UNTRUSTED_PROVENANCE_REPOSITORY}" "${UNTRUSTED_SECRET_USERNAME:-${UNTRUSTED_INPUT_USERNAME}}" "${UNTRUSTED_PASSWORD}" "${GCP_ACCESS_TOKEN}"
```

File: .github/workflows/scripts/registry-login.sh [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+password="${3}"
+gcp_access_token="${4}"
+
+default_registry="docker.io"
+
+# NOTE: Some docker images are of the form <org>/<name>
+# Here we get the first part and check if it has a '.' or ':'
+# character in it to see if it's a domain name.
+# See: https://stackoverflow.com/questions/37861791/how-are-docker-image-names-parsed#37867949
+maybe_domain=$(echo "${image_or_repo}" | cut -f1 -d "/" | grep -E "\.|:" || true)
```
Comment (ramonpetgrave64) [inline .github/workflows/scripts/registry-login.sh:29]: I think the your new behavior would be equivalent, but let's copy-paste to be sure.
```suggestion
maybe_domain=$(echo "${image_or_repo}" | cut -f1 -d "/" | { grep -E "\.|:" || true; })
```

**Pull request closed without merging**